### PR TITLE
Add runnable for rust main function

### PR DESCRIPTION
**Pull request by panghu-huang**
Release Notes:

- N/A

https://github.com/zed-industries/zed/assets/21101490/7a57805c-1d31-48b2-bc2c-3a6f0b730d72

### CODE REVIEW

File: crates/languages/src/rust.rs [hunk 1]
```diff
@@ -426,6 +426,20 @@ impl ContextProvider for RustContextProvider {
...
+            TaskTemplate {
+                label: format!(
+                    "cargo run -p {}",
+                    RUST_PACKAGE_TASK_VARIABLE.template_value()
+                ),
+                command: "cargo".into(),
+                args: vec![
+                    "run".into(),
+                    "-p".into(),
+                    RUST_PACKAGE_TASK_VARIABLE.template_value(),
```
Comment (osiewicz) [inline crates/languages/src/rust.rs:438]: Nit: we should also pass a name of the current file without file extension, as there may be multiple bin target for a single package;
tl;dr: Instead of `cargo run -p RUST_PACKAGE_TASK_VARIABLE`, I'd do `cargo run -p RUST_PACKAGE_TASK_VARIABLE --bin ZED_STEM`

Comment (panghu-huang) [reply crates/languages/src/rust.rs:450]: Thank you. I indeed overlooked that point and have updated it accordingly.

```bash
cargo run -p RUST_PACKAGE_TASK_VARIABLE --bin ZED_CUSTOM_RUST_BIN_NAME
```

File: crates/languages/src/rust.rs [hunk 2]
```diff
@@ -426,6 +437,23 @@ impl ContextProvider for RustContextProvider {
...
+                    RUST_BIN_NAME_TASK_VARIABLE.template_value(),
+                ),
+                command: "cargo".into(),
+                args: vec![
+                    "run".into(),
+                    "-p".into(),
+                    RUST_PACKAGE_TASK_VARIABLE.template_value(),
+                    "--bin".into(),
+                    RUST_BIN_NAME_TASK_VARIABLE.template_value(),
```
Comment (osiewicz) [inline crates/languages/src/rust.rs:452]: Do we need `RUST_BIN_NAME_TASK_VARIABLE` even? I think you could just use `ZED_STEM`; `cargo metadata` can take a while, and context evaluation should be as quick as possible.

Comment (panghu-huang) [reply crates/languages/src/rust.rs:452]: Yes, cargo run does not support specifying a bin using a path. That's why it's necessary to retrieve the bin name from the metadata.

```shell
❯ cargo run --help

Usage: cargo run [OPTIONS] [ARGS]...

Target Selection:
      --bin [<NAME>]      Name of the bin target to run
```

Comment (osiewicz) [reply crates/languages/src/rust.rs:452]: `ZED_STEM` is not the full path though; it's just the name of the file, without a `.rs` extension. AFAIK there's no way to have a bin with a name different than it's stem, so IMHO it'd be better to use stem instead of invoking `cargo metadata`.

Comment (jansol) [reply crates/languages/src/rust.rs:452]: You can do [e.g. this](https://github.com/jansol/texpresso/blob/de748cd341a8e14ea0a89070469e09588ca720bd/cli/Cargo.toml#L14), if I understood the question correctly?

Comment (panghu-huang) [reply crates/languages/src/rust.rs:452]: Hi @osiewicz @jansol 

Currently, the discussion is about the need to specify the bin target configured in the `Cargo.toml` when running `cargo run`.

For example, with an additional bin configuration in `Cargo.toml`:

```toml
[[bin]]
name = "bin-name"
path = "src/bin.rs"
```

To run the main function in `src/bin.rs`, we need to use `cargo run --bin bin-name`, otherwise, an error will occur.

截屏2024-06-18 08 20 24

Comment (osiewicz) [reply crates/languages/src/rust.rs:452]: Alright, thanks you both for clearing that up and I'm sorry; I was wrong. 
I wonder though if we could avoid having to run `cargo metadata` at all when there's no main function in the current function; as is, we'll do it for each time we create task context for Rust.

File: crates/languages/src/rust/runnables.scm [hunk 3]
```diff
@@ -25,3 +25,15 @@
...
     (#set! tag rust-test)
 )
+
+; Rust main function
+(
+    (
+        (function_item
+            name: (_) @run
+            body: _
+        ) @end
```
Comment (osiewicz) [inline crates/languages/src/rust/runnables.scm:35]:
```suggestion
        ) @_rust_main_function_end
```

To continue thought from https://github.com/zed-industries/zed/pull/13087/files#r1644234771,
we could give this capture a distinct name and check whether it's set in `TaskContext` passed into `build_context`; if it is, that means that there's a main function in the current file. If not, we can avoid running `cargo metadata` altogether.

Comment (panghu-huang) [reply crates/languages/src/rust/runnables.scm]: @osiewicz  Thank you for the suggestion, I've updated the implementation as follows:
1. If it's a Rust main function, the package ID and bin target name are retrieved from `cargo metadata`.
2. If it's not a Rust main function, the package ID is retrieved from `cargo pkgid`.

**Review by osiewicz — APPROVED** [review]
LGTM, thanks @panghu-huang (and @jansol) :)

**Comment by panghu-huang** [comment]
> @panghu-huang : What happens if `cargo` is not installed? Does the message `cargo run -p {} --bin {}` still appear?

If the command execution fails or the corresponding bin name is not found, clicking the run button will have no effect.

**Comment by osiewicz** [comment]
@Angelk90 this feels a bit like nitpicking and is not an issue with this PR per se. Please open a separate ticket if surfacing tasks that may not be applicable to a given development environment is bugging you.
